### ISSUE TITLE
[Protocol3] Verifier v3 zkSNARK bug

### ISSUE THREAD

**Issue by HarryR** [Loopring/protocols#309]
Please ensure that all inputs you pass to the verifier are within the SNARK scalar field.

In the following locations:

 * https://github.com/Loopring/protocols/blob/master/packages/loopring_v3/contracts/thirdparty/Verifier.sol#L121
 * https://github.com/Loopring/protocols/blob/master/packages/loopring_v3/contracts/thirdparty/BatchVerifier.sol#L67
 * https://github.com/Loopring/protocols/blob/master/packages/loopring_v3/contracts/thirdparty/Pairing.sol#L87
 * https://github.com/Loopring/protocols/blob/master/packages/loopring_v3/contracts/thirdparty/BatchVerifier.sol#L86

For example:

```solidity
require( mul_input > 0 && mul_input < 21888242871839275222246405745257275088548364400416034343698204186575808495617 ); 
```

Where `q=21888242871839275222246405745257275088548364400416034343698204186575808495617` is the scalar field for zkSNARK inputs.

Not doing so could allow input aliasing, where the resulting point used as input is the same but the on-chain value is different.

For example, if you have a 'nullifier' which prevents double spend, the following piece of code can find alternative inputs which (when reduced modulo `q`) would result in the same value:

```python
p = 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
x = 0xb3e98c3f7346edc7a04b2abd18ca16c11c9419c3c50bd1a34cc485808e895d8e
y = x % p
import z3
s = z3.Solver()
k = z3.BitVec('k', 256)
s.add(k != x)
s.add(k % p == y)
while s.check():
    model = s.model()
    found = int(str(model[k]))
    print(found)
    s.add(k != found)
```

**Comment by Brechtpd** [comment]
Thanks for this @HarryR ! This should have been fixed in #310.

Just to understand this fully, the way we use the verifier _currently_ (hashing the data with `sha256` and only this on-chain calculated input is used to verify the blocks) makes it very hard/impossible to actually exploit this right? You would have to find an input that is still valid (well, at least for the values in the input checked on-chain) AND generate a `sha256` hash which aliases the correct one in the snark field (which you could argue is still the case with the fix because to make the hash fit in the field we have to reduce the length of the hash which also produces potential aliases I guess).

I also added some extra checks for the Poseidon hash inputs because you have the same problem here. I was actually aware of the danger here and the extra checks should not be needed with how we currently hash the values (the critical values are already forced to be smaller than the max field value) but I added some extra just to be safe in the future.

**Comment by HarryR** [comment]
Hi

 Yea absolutely it depends on how you use it, and whether or not finding a pre-image would be necessary, so I agree that if that's your case it shouldn't be a problem.

 I think a few extra sanity checks is good, we're just trying to raise more awareness for this kind of issue :D
